aggregator/src/aggregator.rs

```diff
@@ -3297,24 +3296,54 @@ impl VdafOps {
                     // To ensure that concurrent aggregations don't write into a
                     // currently-nonexistent batch aggregation, we write (empty) batch
                     // aggregations for any that have not already been written to storage.
-                    let empty_batch_aggregations = empty_batch_aggregations(
+                    let batch_aggregations_iter = BatchAggregationsIterator::new(
                         &task,
                         batch_aggregation_shard_count,
                         aggregate_share_req.batch_selector().batch_identifier(),
                         &aggregation_param,
-                        &batch_aggregations,
+                        batch_aggregations,
                     );
 
-                    let (mut helper_aggregate_share, report_count, _, checksum) =
-                        compute_aggregate_share::<SEED_SIZE, B, A>(&task, &batch_aggregations)
-                            .await
-                            .map_err(|e| datastore::Error::User(e.into()))?;
+                    let aggregate_share = Arc::new(Mutex::new(AggregateShareComputer::new(&task)));
+
+                    // Rather than awaiting all the futures at once, which can cause heap growth
+                    // proportional to the number of batch aggregations to write, put them into an
+                    // buffered stream so that they get run in groups of bounded size. #3857
+                    futures::stream::iter(
+                        batch_aggregations_iter
+                            .clone()
```

Comment:
This clone appears to be unnecessary. This could be expensive, since the iterator contains a `HashMap`.